HDDS-8778. Support recursive volume delete using Ozone sh command.

ashishkumar50 opened the pull request:
## What changes were proposed in this pull request?

Currently recursive volume delete is supported only using fs command. But fs can be used only for FSO and Legacy bucket. Since volume can contain mix bucket types and if volume contains some OBS bucket in it then fs rm command fails in between. Currently there is no way for user to delete volume in easy way. 
In this PR, As sh can be used for all bucket types, using ozone sh command user can now perform volume delete recursively. 

Description about new the command: 
Create Volume: vol1 and few buckets/keys inside it.

- Volume delete fails as volume is not empty and we are trying to delete without recursive.

```
ozone sh volume delete o3://localhost:9862/vol1
VOLUME_NOT_EMPTY 
```

- Use -r for recursive volume delete. Interactive confirmation before delete, if input is other than "yes" then make operation cancelled.

```
ozone sh volume delete -r -id localhost:9862 o3://localhost:9862/vol1
This command will delete volume recursively.
There is no recovery option after using this command, and no trash for FSO buckets.
Delay is expected running this command.
Enter 'yes' to proceed': no
Operation cancelled.

```

- Interactive confirmation before recursive delete, Enter "yes" then volume should be deleted

```
ozone sh volume delete -r -id localhost:9862 o3://localhost:9862/vol1/
This command will delete volume recursively.
There is no recovery option after using this command, and no trash for FSO buckets.
Delay is expected running this command.
Enter 'yes' to proceed': yes
Volume vol1 is deleted
```

- Override interactive confirmation by giving "--yes" while running command itself

```
ozone sh volume delete -r --yes -id localhost:9862 o3://localhost:9862/vol1
Volume vol1 is deleted
```

Apart from these options we can also provide number of threads(-t "no of threads") for deletion.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8778

## How was this patch tested?

Added new integration test and verified.

ashishkumar50 commented:
@sadanand48 @sumitagrawl , Please help to review.

sadanand48 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java:
```diff
@@ -18,27 +18,220 @@
+      final Path path = new Path(ofsPrefix);
+      OzoneConfiguration clientConf = new OzoneConfiguration(getConf());
+      clientConf.set(FS_DEFAULT_NAME_KEY, hostPrefix);
+      clientConf.setInt(FS_TRASH_INTERVAL_KEY, 0);
```
Setting trash interval is not required as we are directly calling fs.delete() and not moveToTrash/rename

ashishkumar50 replied:
Removed FS_TRASH_INTERVAL_KEY set as not required.

sadanand48 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java:
```diff
@@ -18,27 +18,220 @@
+          return;
+        }
+        if (OmUtils.isServiceIdsDefined(getConf()) &&
+            Strings.isNullOrEmpty(omServiceId)) {
```
Should it be OR here ?
```
if (OmUtils.isServiceIdsDefined(getConf()) ||
            Strings.isNullOrEmpty(omServiceId))
```

ashishkumar50 replied:
I think AND is correct, because if serviceIds are configured then we need omserviceId.

sadanand48 replied:
Does OmUtils.isServiceIdsDefined(getConf()) return true only when multiple id' s are defined? I think it would be better to check if the number of id's are more than 1.

sadanand48 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java:
```diff
@@ -18,27 +18,220 @@
+
+    try {
+      // wait until all Buckets are cleaned or exception occurred.
+      while (numberOfBucketsCleaned.get() != totalBucketCount
```
Whats stopping user to create more buckets in between this operation, IIUC, this op doesn't take a global volume lock throughout the recursive volume delete, Say Initially total bucket count = 10  , later I get to create more buckets in b/w as there is some time where the volume/buckets are not locked  , In some time bucket count = 15 , this will exit when deleted buckets = 10 but there are 5 more added. Volume deletes will fail in such a case.

ashishkumar50 replied:
Agree with this, this has been already thought and discussed before. When user has decided to delete volume completely, user need to ensure no keys/buckets are getting created in that volume. If that happens then that is user operation problem and volume delete will fail normally. Even if this happen running same command again will delete volume normally.

sadanand48 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java:
```diff
@@ -18,27 +18,220 @@
+    try {
+      if (!bucket.isLink()) {
+        Iterator<? extends OzoneKey> iterator = bucket.listKeys(null);
+        while (iterator.hasNext()) {
```
Same issue here, if new keys are written by another client concurrently , the iterator reference is older and doesn't have the new keys and bucket delete fails. Ideal way to solve this to take a bucket lock on the server , iterate through all keys and delete. If we are sure that the bucket is to be deleted , taking a bucket lock is ok as it doesn't take a toll on other ops as we know we wont be doing any ops on this bucket.

sadanand48 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java:
```diff
@@ -18,27 +18,220 @@
+      names = {"-skipTrash"},
+      description = "Delete volume without trash"
+  )
+  private boolean bSkipTrash = false;
```
Just a suggestion : Given that the operation is too destructive, can we ask a prompt from the user to type something like yes/no just like some linux ops and make the user aware that this would take some time? This will prevent accidental deletion.

ashishkumar50 replied:
Thanks for suggestion, added.

sadanand48 reviewed (commented):
Thanks @ashishkumar50 for the patch, left few comments.

ashishkumar50 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java:
```diff
@@ -18,27 +18,220 @@
+  }
+
+  private void doCleanBuckets() throws InterruptedException {
+    int threadPoolSize = 10;
```
Made it configurable.

errose28 commented:
@ashishkumar50 Could you include what the CLI would look like in the PR description? What subcommand is used, what flags are supported, is there a confirmation prompt, can it be overridden, etc. These UX aspects are probably what people would have the most comments on.

ashishkumar50 commented:
> @ashishkumar50 Could you include what the CLI would look like in the PR description? What subcommand is used, what flags are supported, is there a confirmation prompt, can it be overridden, etc. These UX aspects are probably what people would have the most comments on.

@errose28, Update PR description with all options.

errose28 commented:
Why is there a -skipTrash option? It seems the same as -r and o3 has no concept of trash.

ashishkumar50 commented:
> Why is there a -skipTrash option? It seems the same as -r and o3 has no concept of trash.

We do support trash using o3 for FSO bucket.
Currently using fs also when we do recursive volume delete it expects -skipTrash option or else it will not allow to delete.
So just to make user aware that once deleted it cannot be recovered by providing this option even for o3.

errose28 commented:
If the command always fails if --skipTrash is not provided then that argument is redundant. I think the confirmation message should specify that there is no trash recovery for FSO buckets when this command is used, then -y  or user input becomes the manual override required.

Also I don't think the recently added `ozone sh key delete --skipTrash` option should be present either. I had a discussion with @swamirishi where this flag seemed to imply that OBS buckets had trash if the flag was not used. I think `ozone sh key delete` should always go to trash on FSO bucket and if --skipTrash is needed, the user should use ofs.

Incorporating these two suggestions together provides a consistent experience across the two commands without redundant arguments does not imply that trash exists where it does not.

sadanand48 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java:
```diff
@@ -18,27 +18,246 @@
+  
+  @CommandLine.Option(names = "-yes",
+      description = "Continue without interactive user confirmation")
+  private boolean yes;
```
You can try using picocli interactive option here to add another prompt, Please check https://picocli.info/#_interactive_password_options or this [stackoverflow answer](https://stackoverflow.com/questions/56438657/implementing-interactive-confirmation-in-picocli)

ashishkumar50 replied:
Thanks for the suggestion, I have updated interactive option using picocli. How does it look like I have updated in PR description.

sadanand48 replied:
Looks great.

ashishkumar50 replied:
I tried using picocli but it seems it has different user experience than other system as you can see Ethan's comment. I checked in other place in ozone code and can see we are using interactive option without picocli. So now I have changed back interactive option without piocli.

ashishkumar50 commented:
> If the command always fails if --skipTrash is not provided then that argument is redundant. I think the confirmation message should specify that there is no trash recovery for FSO buckets when this command is used, then -y or user input becomes the manual override required.
> 
> Also I don't think the recently added `ozone sh key delete --skipTrash` option should be present either. I had a discussion with @swamirishi where this flag seemed to imply that OBS buckets had trash if the flag was not used. I think `ozone sh key delete` should always go to trash on FSO bucket and if --skipTrash is needed, the user should use ofs.
> 
> Incorporating these two suggestions together provides a consistent experience across the two commands without redundant arguments implying that trash exists where it does not.

Removed -skipTrash option, instead showing in interactive description. Updated the change in PR description.

ashishkumar50 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java:
```diff
@@ -18,27 +18,246 @@
+  private boolean bSkipTrash = false;
+  @CommandLine.Option(
+      names = {"-r"},
+      description = "Delete volume recursively"
```
Added in the interactive description when user does recursive delete. About limit number we can have improvement in future for this command. In PR description I have added how the interactive description looks like.

errose28 commented:
> ozone sh volume delete -r=yes -id localhost:9862 o3://localhost:9862/vol2
Volume vol2 is deleted

I think the override should be a separate flag, not an argument to -r. When in doubt, look at existing CLIs for the intuitive thing to do. `rm -r` requires `-f` to bypass the prompt. `pacman` uses `--noconfirm`, `apt` requires `-y` etc. Just a few examples I could think of off the top of my head. Something like `--yes/-y` seems more in line with standards than `-r=yes`.

> ozone sh volume delete -r -id localhost:9862 o3://localhost:9862/vol1
Enter value for -r (This command will delete volume recursively.
There is no trash recovery for FSO buckets using this command.
Delay is expected running this command.
Enter 'yes' to proceed): no
VOLUME_NOT_EMPTY

This should probably also specify that there is no recovery option for OBS buckets either, although we don't have that anyways. Maybe a message like "There is no recovery option after using this command, and no trash for FSO buckets" would help.

Also, is there a follow up PR to implement recursive bucket delete as well?

sadanand48 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java:
```diff
@@ -18,27 +18,240 @@
+  private String omServiceId;
+
+  @CommandLine.Option(names = {"-t", "--threads", "--thread"},
+      description = "Number of threads used to execute")
```
```suggestion
      description = "Number of threads used to execute recursive delete")
```

ashishkumar50 commented:
> > ozone sh volume delete -r=yes -id localhost:9862 o3://localhost:9862/vol2
> > Volume vol2 is deleted
> 
> I think the override should be a separate flag, not an argument to -r. When in doubt, look at existing CLIs for the intuitive thing to do. `rm -r` requires `-f` to bypass the prompt. `pacman` uses `--noconfirm`, `apt` requires `-y` etc. Just a few examples I could think of off the top of my head. Something like `--yes/-y` seems more in line with standards than `-r=yes`.
> 
> > ozone sh volume delete -r -id localhost:9862 o3://localhost:9862/vol1
> > Enter value for -r (This command will delete volume recursively.
> > There is no trash recovery for FSO buckets using this command.
> > Delay is expected running this command.
> > Enter 'yes' to proceed): no
> > VOLUME_NOT_EMPTY
> 
> This should probably also specify that there is no recovery option for OBS buckets either, although we don't have that anyways. Maybe a message like "There is no recovery option after using this command, and no trash for FSO buckets" would help.
> 
> Also, is there a follow up PR to implement recursive bucket delete as well?

I have updated with separate flag to override.
Updated user interactive message.
Yes I will have a follow up PR for recursive volume delete.

errose28 commented:
CLI options LGTM although I will not have time to review the code changes. Thanks for updating those @ashishkumar50

sadanand48 commented:
Thanks @ashishkumar50 for addressing the comments. The changes look good to me. We should also document the newly made changes regarding ozone sh deletes to get clarity on items like trash. Opened HDDS-8916 for the same.

ashishkumar50 commented:
> Thanks @ashishkumar50 for addressing the comments. The changes look good to me. We should also document the newly made changes regarding ozone sh deletes to get clarity on items like trash. Opened [HDDS-8916](https://issues.apache.org/jira/browse/HDDS-8916) for the same.

Okay @sadanand48, I will document these new changes in [HDDS-8916](https://issues.apache.org/jira/browse/HDDS-8916)
